### PR TITLE
[clang-format] Don't count template template parameter as declaration

### LINKED ISSUE #48746: clang-format formats template template parameters inconsistently for Mozilla/Chromium style
|  |  |
| --- | --- |
| Bugzilla Link | [49402](https://llvm.org/bz49402) |
| Version | 11.0 |
| OS | All |

## Extended Description 
Input:
```
  $ cat test.cc 
  template <template <typename> class A,
            template <typename> class B,
            typename C,
            typename D>
  struct TemplatedClass {};
```
Output for Mozilla style:
```
  $ clang-format --style=Mozilla test.cc
  template<template<typename> class A,
           template<typename>
           class B,
           typename C,
           typename D>
  struct TemplatedClass
  {};
```
Output for Chromium style:
```
  $ clang-format --style=Chromium test.cc
  template <template <typename> class A,
            template <typename>
            class B,
            typename C,
            typename D>
  struct TemplatedClass {};
```
One might have expected "template <typename>" and "class B" to be on the same line, as is the case with the first template template parameter.

### LINKED ISSUE #93793: clang-format: Inserts unwanted newlines in template <typename...> typename class template
Desired format:

```
template <template <typename...> typename Xxx,
          template <typename...> typename Yyy,
          typename... T>
class A
{
};

```

Actual format: 

```
template <template <typename...> typename Xxx,
          template <typename...>
          typename Yyy,
          typename... T>
class A
{
};
```

My .clang-format:

```
BasedOnStyle: Microsoft
ColumnLimit: 0
SortIncludes: Never
BreakBeforeBinaryOperators: All
BinPackArguments: False
BinPackParameters: False
LambdaBodyIndentation: OuterScope
IndentCaseBlocks: true
MaxEmptyLinesToKeep: 3
BraceWrapping:
  AfterNamespace: false
  AfterCaseLabel: true
  BeforeLambdaBody: true

```

Using clang-format 14

### PR DESCRIPTION
**Pull request by rymiel**
In ContinuationIndenter::mustBreak, a break is required between a template declaration and the function/class declaration it applies to, if the template declaration spans multiple lines.

However, this also includes template template parameters, which can cause extra erroneous line breaks in some declarations.

This patch makes template template parameters not be counted as template declarations.

Fixes https://github.com/llvm/llvm-project/issues/93793
Fixes https://github.com/llvm/llvm-project/issues/48746

### CODE REVIEW

**Comment by rymiel** [comment]
Note: I understand my solution of adding a member variable is inelegant, I am of course open to ways on making it better

File: clang/lib/Format/TokenAnnotator.cpp [hunk 1]
```diff
@@ -1269,10 +1269,17 @@ class AnnotatingParser {
...
+      TemplateDeclarationDepth++;
+      if (!parseAngle()) {
+        TemplateDeclarationDepth--;
         return false;
-      if (CurrentToken)
+      }
+      TemplateDeclarationDepth--;
+      if (CurrentToken &&
+          !(TemplateDeclarationDepth > 0 &&
+            CurrentToken->isOneOf(tok::kw_typename, tok::kw_class))) {
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:1280]:
```suggestion
          (TemplateDeclarationDepth == 0 ||
            !CurrentToken->isOneOf(tok::kw_typename, tok::kw_class))) {
```

**Review by HazardyKnusperkeks — APPROVED** [review]

File: clang/unittests/Format/TokenAnnotatorTest.cpp [hunk 2]
```diff
@@ -584,6 +584,23 @@ TEST_F(TokenAnnotatorTest, UnderstandsNonTemplateAngleBrackets) {
...
+TEST_F(TokenAnnotatorTest, UnderstandsTemplateTemplateParameters) {
+  auto Tokens = annotate("template <template <typename...> typename X,\n"
+                         "          template <typename...> typename Y,\n"
+                         "          typename... T>\n"
+                         "class A {};");
+  ASSERT_EQ(Tokens.size(), 28u) << Tokens;
+  EXPECT_TOKEN(Tokens[1], tok::less, TT_TemplateOpener);
+  EXPECT_TOKEN(Tokens[3], tok::less, TT_TemplateOpener);
+  EXPECT_TOKEN(Tokens[6], tok::greater, TT_TemplateCloser);
+  EXPECT_EQ(Tokens[6]->ClosesTemplateDeclaration, 0u);
```
[the remark this answers is not in the document]
Comment (rymiel) [reply clang/unittests/Format/TokenAnnotatorTest.cpp]: Thanks, some existing tests had this pattern so I followed those but EXPECT_TRUE and EXPECT_FALSE make more sense

File: clang/lib/Format/TokenAnnotator.cpp [hunk 3]
```diff
@@ -1269,10 +1269,17 @@ class AnnotatingParser {
...
+      TemplateDeclarationDepth++;
+      if (!parseAngle()) {
+        TemplateDeclarationDepth--;
         return false;
-      if (CurrentToken)
+      }
+      TemplateDeclarationDepth--;
+      if (CurrentToken &&
+          (TemplateDeclarationDepth == 0 ||
+           !CurrentToken->isOneOf(tok::kw_typename, tok::kw_class))) {
```
[the remark this answers is not in the document]
Comment (rymiel) [reply clang/lib/Format/TokenAnnotator.cpp:1280]: If parseAngle fails, wouldn't that cause InTemplateDeclaration to forever be true? Is that not an issue?

[the remark this answers is not in the document]
Comment (owenca) [reply clang/lib/Format/TokenAnnotator.cpp:1280]: > If parseAngle fails, wouldn't that cause InTemplateDeclaration to forever be true? Is that not an issue?

If `parseAngle()` returns false, it probably means that the template decl is ill-formed. If so, it doesn't matter if `InTemplateDeclaration` is set to false?

An alternative solution:
```
-  bool parseTemplateDeclaration() {
-    if (CurrentToken && CurrentToken->is(tok::less)) {
-      CurrentToken->setType(TT_TemplateOpener);
-      next();
-      if (!parseAngle())
-        return false;
-      if (CurrentToken)
-        CurrentToken->Previous->ClosesTemplateDeclaration = true;
-      return true;
-    }
-    return false;
+  bool parseTemplateDeclaration(bool InTemplateParameter) {
+    if (!CurrentToken || CurrentToken->isNot(tok::less))
+      return false;
+
+    InTemplateDeclaration = true;
+    CurrentToken->setType(TT_TemplateOpener);
+    next();
+
+    const bool WellFormed = parseAngle();
+    InTemplateDeclaration = InTemplateParameter;
+
+    if (!WellFormed)
+      return false;
+
+    if (CurrentToken && !InTemplateParameter)
+      CurrentToken->Previous->ClosesTemplateDeclaration = true;
+
+    return true;
   }
```

[the remark this answers is not in the document]
Comment (owenca) [reply clang/lib/Format/TokenAnnotator.cpp:1280]: Actually, checking the nesting level of the template brackets like you did is easier to understand. You just don't need to check the type-parameter keys IMO. Below is a rewrite of the function based on your solution:
```cpp
$ git diff
diff --git a/clang/lib/Format/TokenAnnotator.cpp b/clang/lib/Format/TokenAnnotator.cpp
index 958b46c535a9..9be5c09a32dd 100644
--- a/clang/lib/Format/TokenAnnotator.cpp
+++ b/clang/lib/Format/TokenAnnotator.cpp
@@ -1266,23 +1266,22 @@ private:
   }
 
   bool parseTemplateDeclaration() {
-    if (CurrentToken && CurrentToken->is(tok::less)) {
-      CurrentToken->setType(TT_TemplateOpener);
-      next();
-      TemplateDeclarationDepth++;
-      if (!parseAngle()) {
-        TemplateDeclarationDepth--;
-        return false;
-      }
-      TemplateDeclarationDepth--;
-      if (CurrentToken &&
-          (TemplateDeclarationDepth == 0 ||
-           !CurrentToken->isOneOf(tok::kw_typename, tok::kw_class))) {
-        CurrentToken->Previous->ClosesTemplateDeclaration = true;
-      }
-      return true;
-    }
-    return false;
+    if (!CurrentToken || CurrentToken->isNot(tok::less))
+      return false;
+
+    CurrentToken->setType(TT_TemplateOpener);
+    next();
+
+    TemplateDeclarationDepth++;
+    const bool WellFormed = parseAngle();
+    TemplateDeclarationDepth--;
+    if (!WellFormed)
+      return false;
+
+    if (CurrentToken && TemplateDeclarationDepth == 0)
+      CurrentToken->Previous->ClosesTemplateDeclaration = true;
+
+    return true;
   }
 
   bool consumeToken() {
$ 
```

[the remark this answers is not in the document]
Comment (rymiel) [reply clang/lib/Format/TokenAnnotator.cpp]: Thanks!

**Review by owenca — APPROVED** [review]

**Pull request merged**